packages/styled-components/src/models/ServerStyleSheet.js

```diff
@@ -1,96 +1,57 @@
 // @flow
 /* eslint-disable no-underscore-dangle */
+
 import React from 'react';
 import stream from 'stream';
+import { Sheet } from 'styled-sheet';
 
-import { IS_BROWSER, SC_STREAM_ATTR } from '../constants';
+import { IS_BROWSER } from '../constants';
 import StyledError from '../utils/error';
-import StyleSheet from './StyleSheet';
 import StyleSheetManager from './StyleSheetManager';
 
 declare var __SERVER__: boolean;
 
 export default class ServerStyleSheet {
-  instance: StyleSheet;
-
-  masterSheet: StyleSheet;
+  sheet: Sheet;
 
   sealed: boolean;
 
   constructor() {
-    /* The master sheet might be reset, so keep a reference here */
-    this.masterSheet = StyleSheet.master;
-    this.instance = this.masterSheet.clone();
+    this.sheet = new Sheet(undefined, true);
     this.sealed = false;
   }
 
-  /**
-   * Mark the ServerStyleSheet as being fully emitted and manually GC it from the
-   * StyleSheet singleton.
-   */
-  seal() {
-    if (!this.sealed) {
-      /* Remove sealed StyleSheets from the master sheet */
-      const index = this.masterSheet.clones.indexOf(this.instance);
-      this.masterSheet.clones.splice(index, 1);
-      this.sealed = true;
-    }
-  }
-
   collectStyles(children: any) {
     if (this.sealed) {
       throw new StyledError(2);
     }
 
-    return <StyleSheetManager sheet={this.instance}>{children}</StyleSheetManager>;
+    return <StyleSheetManager sheet={this.sheet}>{children}</StyleSheetManager>;
   }
 
   getStyleTags(): string {
-    this.seal();
-    return this.instance.toHTML();
-  }
-
-  getStyleElement() {
-    this.seal();
-    return this.instance.toReactElements();
+    return this.sheet.toHTML();
   }
 
   interleaveWithNodeStream(readableStream: stream.Readable) {
     if (!__SERVER__ || IS_BROWSER) {
       throw new StyledError(3);
     }
 
-    /* the tag index keeps track of which tags have already been emitted */
-    const { instance } = this;
-    let instanceTagIndex = 0;
-
-    const streamAttr = `${SC_STREAM_ATTR}="true"`;
+    this.sealed = true;
+    const { sheet } = this;
 
     const transformer = new stream.Transform({
       transform: function appendStyleChunks(chunk, /* encoding */ _, callback) {
-        const { tags } = instance;
-        let html = '';
-
-        /* retrieve html for each new style tag */
-        for (; instanceTagIndex < tags.length; instanceTagIndex += 1) {
-          const tag = tags[instanceTagIndex];
-          html += tag.toHTML(streamAttr);
-        }
-
-        /* force our StyleSheets to emit entirely new tags */
-        instance.sealAllTags();
-
-        /* prepend style html to chunk */
+        const html = sheet.toHTML();
+        // Force Sheet to "forget" about previously emitted CSS
+        sheet.reset();
```

Comment:
Since there should be a new `sheet` on every request this seems like unnecessary work, shouldn't the entire sheet be garbage collected once the request is done?

Reply:
This is the code inbetween chunks on the SSR stream. We need to ensure that previously emitted CSS is not emitted again. Since there's no "useful" information in a GroupedTag apart from just CSS, the Stylesheet can just throw it away and create a new one